cl/sentinel/gossip.go

```diff
@@ -285,30 +289,55 @@ type GossipSubscription struct {
 	host         peer.ID
 	ch           chan *GossipMessage
 	ctx          context.Context
+	expiration   atomic.Value // Unix nano for how much we should listen to this topic
+	subscribed   atomic.Bool
 
 	topic *pubsub.Topic
 	sub   *pubsub.Subscription
 
 	cf context.CancelFunc
 	rf pubsub.RelayCancelFunc
 
-	setup  sync.Once
 	stopCh chan struct{}
 }
 
-func (sub *GossipSubscription) Listen() (err error) {
-	sub.setup.Do(func() {
-		sub.stopCh = make(chan struct{}, 3)
-		sub.sub, err = sub.topic.Subscribe()
-		if err != nil {
-			err = fmt.Errorf("failed to begin topic %s subscription, err=%w", sub.topic.String(), err)
-			return
+func (sub *GossipSubscription) Listen() {
+	go func() {
+		var err error
+		checkingInterval := time.NewTicker(100 * time.Millisecond)
+		for {
+			select {
+			case <-sub.ctx.Done():
+				return
+			case <-checkingInterval.C:
+				expirationTime := sub.expiration.Load().(time.Time)
+				if sub.subscribed.Load() && time.Until(expirationTime) < 0 {
```

Comment:
how about to use `time.After(expirationTime)`?

Reply:
what if the expiry is renewed?